### PR TITLE
Replaced Convert Based On ToASCII With Version Based on Export External Contribution

### PR DESCRIPTION
**Pull request by andrew-lunarg**
It should do what the old Convert did, just with a slightly different schema, some nice new features, and based on a more extensible foundation of the Modern C++ JSON Library. It handles the union of the block types previously handled by the two tools, so state markers, annotations, and metacommands as well as functions and methods.
D3D12 support is still there with the old schema hidden behind an experimental build flag.
The many commits may be reordered and will be grouped and squashed into a smaller number befor merge to dev after feedback.

```
./dbuild/linux/x64/cmake_output/tools/convert/gfxrecon-convert --help

gfxrecon-convert - A tool to convert the contents of GFXReconstruct capture files to JSON.

Usage:
  gfxrecon-convert [-h | --help] [--version] <file>

Required arguments:
  <file>		Path to the GFXReconstruct capture file to be converted
        		to text.

Optional arguments:
  -h			Print usage information and exit (same as --help).
  --version		Print version information and exit.
  --output file		'stdout' or a path to a file to write JSON output
        		to. Default is the input filepath with "gfxr" replaced by "json".
  --format <format>	JSON format to write.
           json		Standard JSON format (indented)
           jsonl	JSON lines format (every object in a single line)
  --include-binaries	Dump binaries from Vulkan traces in a separate file with an unique name. The main JSON file
                    	will include a reference with the file name. The binary files are dumped in a subdirectory
  --expand-flags	Print flags values from Vulkan traces with its correspondent symbolic representation. Otherwise,
                	the flags are printed as hexadecimal value.
  --file-per-frame	Creates a new file for every frame processed. Frame number is added as a suffix
                  	to the output file name.
```

`./dbuild/linux/x64/cmake_output/tools/convert/gfxrecon-convert --output stdout raspberrypi/vkcube_rpi.gfxr`:
```json
[
{
  "header": {
    "source-path": "/depot/gfxr-traces/from_nas/smb_traces/vulkan/GFXR/raspberrypi/vkcube_rpi.gfxr",
    "gfxrecon-version": "0.9.19-dev (andy-export-tool-07-changes:e2e7c6a*)",
    "vulkan-version": "1.3.243"
  }
},
{
  "index": 0,
  "function": {
    "name": "vkCreateInstance",
    "thread": 1,
    "return": "VK_SUCCESS",
    "args": {
      "pCreateInfo": {
        "sType": "VK_STRUCTURE_TYPE_INSTANCE_CREATE_INFO",
        "flags": "0x00000001",
        "pApplicationInfo": {
          "sType": "VK_STRUCTURE_TYPE_APPLICATION_INFO",
          "pApplicationName": "vkcube",
          "applicationVersion": 0,
          "pEngineName": "vkcube",
          "engineVersion": 0,
          "apiVersion": 4194304,
          "pNext": null
        },
        "enabledLayerCount": 0,
        "ppEnabledLayerNames": null,
        "enabledExtensionCount": 4,
        "ppEnabledExtensionNames": [
          "VK_KHR_get_physical_device_properties2",
          "VK_KHR_surface",
          "VK_KHR_xcb_surface",
          "VK_KHR_portability_enumeration"
        ],
        // pNext reordered to bottom of object:
        "pNext": null
      },
      "pAllocator": null,
      "pInstance": 1
    }
  }
},
...
{
  "index": 8,
  "function": {
    "name": "vkGetPhysicalDeviceProperties",
    "thread": 1,
    "args": {
      "physicalDevice": 2,
      "pProperties": {
        "apiVersion": 4202720,
        "driverVersion": 92282979,
        "vendorID": 5348,
        "deviceID": 3192414163,
        "deviceType": "VK_PHYSICAL_DEVICE_TYPE_INTEGRATED_GPU",
        "deviceName": "V3D 4.2",
        "pipelineCacheUUID": "5e9123edbd30bc930c69bd76cca56b",
        "limits": {
          "maxImageDimension1D": 4096,
...
```

`./dbuild/linux/x64/cmake_output/tools/convert/gfxrecon-convert --output stdout --expand-flags raspberrypi/vkcube_rpi.gfxr`:
```json
{
  "index": 28,
  "function": {
    "name": "vkGetPhysicalDeviceMemoryProperties",
    "thread": 1,
    "args": {
      "physicalDevice": 2,
      "pMemoryProperties": {
        "memoryTypeCount": 1,
        "memoryTypes": [
          {
            // Pretty flags:
            "propertyFlags": "VK_MEMORY_PROPERTY_DEVICE_LOCAL_BIT|VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT|VK_MEMORY_PROPERTY_HOST_COHERENT_BIT",
            "heapIndex": 0
          }
        ],
        "memoryHeapCount": 1,
        "memoryHeaps": [
          {
            "size": 969572352,
            "flags": "VK_MEMORY_HEAP_DEVICE_LOCAL_BIT"
          }
        ]
      }
    }
  }
},
...
{
  "index": 83,
  "function": {
    "name": "vkCreateShaderModule",
    "thread": 1,
    "return": "VK_SUCCESS",
    "args": {
      "device": 4,
      "pCreateInfo": {
        "sType": "VK_STRUCTURE_TYPE_SHADER_MODULE_CREATE_INFO",
        "flags": "0x00000000",
        "codeSize": 1280,
        // Host pointer replaced with this tag rather than the captured host address:
        "pCode": "[Binary data]",
        "pNext": null
      },
      "pAllocator": null,
      "pShaderModule": 40
    }
  }
},
{
  "index": 84,
  "function": {
    "name": "vkCreatePipelineCache",
    "thread": 1,
    "return": "VK_SUCCESS",
    "args": {
      "device": 4,
      "pCreateInfo": {
        "sType": "VK_STRUCTURE_TYPE_PIPELINE_CACHE_CREATE_INFO",
        "flags": "0x00000000",
        "initialDataSize": 0,
        // Host pointer replaced with this tag rather than the captured host address:
        "pInitialData": "[Binary data]",
        "pNext": null
      },
      "pAllocator": null,
      "pPipelineCache": 41,
        // Host pointer replaced with this tag rather than the captured host address:
      "pData": "[Binary data]"
    }
  }
},
{
  "index": 85,
  "function": {
    "name": "vkCreateGraphicsPipelines",
    "thread": 1,
    "return": "VK_SUCCESS",
    "args": {
      "device": 4,
      "pipelineCache": 41,
      "createInfoCount": 1,
      "pCreateInfos": [
        {
          "sType": "VK_STRUCTURE_TYPE_GRAPHICS_PIPELINE_CREATE_INFO",
          "flags": "0x00000000",
          "stageCount": 2,
          "pStages": [
            {
              "sType": "VK_STRUCTURE_TYPE_PIPELINE_SHADER_STAGE_CREATE_INFO",
              "flags": "0x00000000",
              "stage": "VK_SHADER_STAGE_VERTEX_BIT",
              "module": 39,
              "pName": "main",
              "pSpecializationInfo": null,
              "pNext": null
            },
            {
              "sType": "VK_STRUCTURE_TYPE_PIPELINE_SHADER_STAGE_CREATE_INFO",
              "flags": "0x00000000",
              "stage": "VK_SHADER_STAGE_FRAGMENT_BIT",
              "module": 40,
              "pName": "main",
              "pSpecializationInfo": null,
              "pNext": null
            }
          ],
          "pVertexInputState": {
            "sType": "VK_STRUCTURE_TYPE_PIPELINE_VERTEX_INPUT_STATE_CREATE_INFO",
            "flags": "0x00000000",
            "vertexBindingDescriptionCount": 0,
            "pVertexBindingDescriptions": null,
            "vertexAttributeDescriptionCount": 0,
            "pVertexAttributeDescriptions": null,
            "pNext": null
          },
          "pInputAssemblyState": {
            "sType": "VK_STRUCTURE_TYPE_PIPELINE_INPUT_ASSEMBLY_STATE_CREATE_INFO",
            "flags": "0x00000000",
            "topology": "VK_PRIMITIVE_TOPOLOGY_TRIANGLE_LIST",
            "primitiveRestartEnable": false,
            "pNext": null
          },
          "pTessellationState": null,
          "pViewportState": {
            "sType": "VK_STRUCTURE_TYPE_PIPELINE_VIEWPORT_STATE_CREATE_INFO",
            "flags": "0x00000000",
            "viewportCount": 1,
            "pViewports": null,
            "scissorCount": 1,
            "pScissors": null,
            "pNext": null
          },
          "pRasterizationState": {
            "sType": "VK_STRUCTURE_TYPE_PIPELINE_RASTERIZATION_STATE_CREATE_INFO",
            "flags": "0x00000000",
            "depthClampEnable": false,
            "rasterizerDiscardEnable": false,
            "polygonMode": "VK_POLYGON_MODE_FILL",
            "cullMode": "VK_CULL_MODE_BACK_BIT",
            "frontFace": "VK_FRONT_FACE_COUNTER_CLOCKWISE",
            "depthBiasEnable": false,
            "depthBiasConstantFactor": 0.0,
            "depthBiasClamp": 0.0,
            "depthBiasSlopeFactor": 0.0,
            "lineWidth": 1.0,
            "pNext": null
          },
          "pMultisampleState": {
            "sType": "VK_STRUCTURE_TYPE_PIPELINE_MULTISAMPLE_STATE_CREATE_INFO",
            "flags": "0x00000000",
            "rasterizationSamples": "VK_SAMPLE_COUNT_1_BIT",
            "sampleShadingEnable": false,
            "minSampleShading": 0.0,
            "pSampleMask": null,
            "alphaToCoverageEnable": false,
            "alphaToOneEnable": false,
            "pNext": null
          },
          "pDepthStencilState": {
            "sType": "VK_STRUCTURE_TYPE_PIPELINE_DEPTH_STENCIL_STATE_CREATE_INFO",
            "flags": "0x00000000",
            "depthTestEnable": true,
            "depthWriteEnable": true,
            "depthCompareOp": "VK_COMPARE_OP_LESS_OR_EQUAL",
            "depthBoundsTestEnable": false,
            "stencilTestEnable": false,
            "front": {
              "failOp": "VK_STENCIL_OP_KEEP",
              "passOp": "VK_STENCIL_OP_KEEP",
              "depthFailOp": "VK_STENCIL_OP_KEEP",
              "compareOp": "VK_COMPARE_OP_ALWAYS",
              "compareMask": 0,
              "writeMask": 0,
              "reference": 0
            },
            "back": {
              "failOp": "VK_STENCIL_OP_KEEP",
              "passOp": "VK_STENCIL_OP_KEEP",
              "depthFailOp": "VK_STENCIL_OP_KEEP",
              "compareOp": "VK_COMPARE_OP_ALWAYS",
              "compareMask": 0,
              "writeMask": 0,
              "reference": 0
            },
            "minDepthBounds": 0.0,
            "maxDepthBounds": 0.0,
            "pNext": null
          },
          "pColorBlendState": {
            "sType": "VK_STRUCTURE_TYPE_PIPELINE_COLOR_BLEND_STATE_CREATE_INFO",
            "flags": "0x00000000",
            "logicOpEnable": false,
            "logicOp": "VK_LOGIC_OP_CLEAR",
            "attachmentCount": 1,
            "pAttachments": [
              {
                "blendEnable": false,
                "srcColorBlendFactor": "VK_BLEND_FACTOR_ZERO",
                "dstColorBlendFactor": "VK_BLEND_FACTOR_ZERO",
                "colorBlendOp": "VK_BLEND_OP_ADD",
                "srcAlphaBlendFactor": "VK_BLEND_FACTOR_ZERO",
                "dstAlphaBlendFactor": "VK_BLEND_FACTOR_ZERO",
                "alphaBlendOp": "VK_BLEND_OP_ADD",
                // Pretty flags:
                "colorWriteMask": "VK_COLOR_COMPONENT_R_BIT|VK_COLOR_COMPONENT_G_BIT|VK_COLOR_COMPONENT_B_BIT|VK_COLOR_COMPONENT_A_BIT"
              }
            ],
            "blendConstants": [
              0.0,
              0.0,
              0.0,
              0.0
            ],
            "pNext": null
          },
          "pDynamicState": {
            "sType": "VK_STRUCTURE_TYPE_PIPELINE_DYNAMIC_STATE_CREATE_INFO",
            "flags": "0x00000000",
            "dynamicStateCount": 2,
            "pDynamicStates": [
              "VK_DYNAMIC_STATE_VIEWPORT",
              "VK_DYNAMIC_STATE_SCISSOR"
            ],
            "pNext": null
          },
          "layout": 37,
          "renderPass": 38,
          "subpass": 0,
          "basePipelineHandle": 0,
          "basePipelineIndex": 0,
          "pNext": null
        }
      ],
      "pAllocator": null,
      "pPipelines": [
        42
      ]
    }
  }
},
...
{
  "index": 1984,
  "function": {
    "name": "vkDestroyInstance",
    "thread": 1,
    "args": {
      "instance": 1,
      "pAllocator": null
    }
  }
}
```

**JSON Lines:**

`./dbuild/linux/x64/cmake_output/tools/convert/gfxrecon-convert --format jsonl --output stdout --expand-flags raspberrypi/vkcube_rpi.gfxr`:
```json
{"header":{"source-path":"/depot/gfxr-traces/from_nas/smb_traces/vulkan/GFXR/raspberrypi/vkcube_rpi.gfxr","gfxrecon-version":"0.9.19-dev (andy-export-tool-07-changes:e2e7c6a*)","vulkan-version":"1.3.243"}}
{"index":0,"function":{"name":"vkCreateInstance","thread":1,"return":"VK_SUCCESS","args":{"pCreateInfo":{"sType":"VK_STRUCTURE_TYPE_INSTANCE_CREATE_INFO","flags":                                                 "VK_INSTANCE_CREATE_ENUMERATE_PORTABILITY_BIT_KHR","pApplicationInfo":{"sType":"VK_STRUCTURE_TYPE_APPLICATION_INFO","pApplicationName":"vkcube","applicationVersion":0,"pEngineName":"vkcube","engineVersion":0,   "apiVersion":4194304,"pNext":null},"enabledLayerCount":0,"ppEnabledLayerNames":null,"enabledExtensionCount":4,"ppEnabledExtensionNames":["VK_KHR_get_physical_device_properties2","VK_KHR_surface",                "VK_KHR_xcb_surface","VK_KHR_portability_enumeration"],"pNext":null},"pAllocator":null,"pInstance":1}}}
{"index":1,"function":{"name":"vkEnumeratePhysicalDevices","thread":1,"return":"VK_SUCCESS","args":{"instance":1,"pPhysicalDeviceCount":1,"pPhysicalDevices":null}}}
{"index":2,"function":{"name":"vkEnumeratePhysicalDevices","thread":1,"return":"VK_SUCCESS","args":{"instance":1,"pPhysicalDeviceCount":1,"pPhysicalDevices":[2]}}}
// Metablocks don't have their indexes output yet but that will be added in the same position as currently used for functions later:
{"meta":{"name":"SetDevicePropertiesCommand","args":{"physical_device_id":2,"api_version":4202720,"driver_version":92282979,"vendor_id":5348,"device_id":3192414163,"device_type":1,"pipeline_cache_uuid":         "5e9123edbd30bc930c69bd76cca56b","device_name":"V3D 4.2"}}}
{"meta":{"name":"SetDeviceMemoryPropertiesCommand","args":{"physical_device_id":2}}}
...
```

### CODE REVIEW

File: framework/decode/custom_vulkan_struct_decoders_to_string.cpp
Comment (andrew-lunarg) [inline framework/decode/custom_vulkan_struct_decoders_to_string.cpp:1]: I killed it, I killed it all!

File: framework/util/to_string.h [hunk 1]
```diff
@@ -28,10 +28,13 @@
```
Comment (andrew-lunarg) [inline framework/util/to_string.h:28]: Remaining to_string code is used by DX12 ascii consumer and in some logging code that wants a pretty representation of a value it wants to show to the user. There may still be dead code lurking that we can continue to cull when we spot it later.

**Comment by andrew-lunarg** [comment]
> CI gfxreconstruct build # 2602 failed.

Manual build #2603 passed.

**Review by bradgrantham-lunarg — APPROVED** [review]
I noted a misspelling.

Otherwise looks good except a couple of copyrights were changed or lost.  This PR may need a pass to verify copyrights have been updated correctly.

**Pull request merged**